File: Medium/Valid-IP-Address/validIpAddress.ts
// Given a string with digits, find out all valid ip addresses that could be formed
// by separating the string with dot delimiter
// O(1) Time | O(1) Space; Reason: both time and space of possible ip address has a upper bound O(2^32),
// regardless of what input is provided
const ip = '1921680';

// helper func to check for 2 invalid cases:
// 1. whether the part (substring) is non-zero numbers padded with leading zero
// 2. whether the integer is greater than 255
const isValidPart = (part: string): boolean => {
  const numPart = parseInt(part);
  if (numPart > 255) return false;

  return numPart.toString().length === part.length;
};

const getValidIpList = (ipStr: string): string[] => {
  const ipList: string[] = [];
  let currentIP = '';
  let firstIdx: number = 1,
    secondIdx: number,
    thirdIdx: number;
  let first: string, second: string, third: string, fourth: string;

  if (ipStr.startsWith('0')) return [];

  // while (firstIdx < 4 && firstIdx < ipStr.length - 2) {
  while (firstIdx < Math.min(4, ipStr.length - 2)) {
    first = ipStr.slice(0, firstIdx);
    if (!isValidPart(first)) break;

    secondIdx = firstIdx + 1;
    thirdIdx = secondIdx + 1;
    // while (secondIdx < 7 && secondIdx < ipStr.length - 1) {
    while (secondIdx < Math.min(7, ipStr.length - 1)) {
      second = ipStr.slice(firstIdx, secondIdx);
      if (!isValidPart(second)) break;

      thirdIdx = secondIdx + 1;
      while (thirdIdx < ipStr.length) {
        third = ipStr.slice(secondIdx, thirdIdx);
        fourth = ipStr.slice(thirdIdx);

        if (isValidPart(third) && isValidPart(fourth)) {
          currentIP = [first, second, third, fourth].join('.');
          ipList.push(currentIP);
        }
        thirdIdx++;
      }
      secondIdx++;
    }
    firstIdx++;
  }

  return ipList;
};

console.log(getValidIpList(ip));
console.log(getValidIpList('2125074'));
console.log(getValidIpList('11112'));
console.log(getValidIpList('2553072'));
console.log(getValidIpList('035345'));

File: Strings/Valid-IP-Address/ValidIpAddress.go
// Given a string with digits, find out all valid ip addresses that could be formed
// by separating the string with dot delimiter
// O(1) Time | O(1) Space; Reason: both time and space of possible ip address has a upper bound O(2^32),
// regardless of what input is provided
package main

import (
	"fmt"
	"math"
	"strconv"
)

const ip = "1921680"

func main() {
	fmt.Printf("%#v\n", getValidIpList(ip))
	fmt.Printf("%#v\n", getValidIpList("2125074"))
	fmt.Printf("%#v\n", getValidIpList("11112"))
	fmt.Printf("%#v\n", getValidIpList("2553072"))
	fmt.Printf("%#v\n", getValidIpList("035345"))
}

func isValidPart(part string) bool {
	numPart, err := strconv.ParseInt(part, 10, 64)
	if err != nil {
		fmt.Println(err)
		return false
	}

	if numPart > 255 {
		return false
	}
	return len(fmt.Sprint(numPart)) == len(part)
}

func getValidIpList(ipStr string) []string {
	var ipList []string
	var currentIP string
	firstIdx, secondIdx, thirdIdx := 1, 0, 0
	var first, second, third, fourth string

	if string(ipStr[0]) == "0" {
		return ipList
	}

	for firstIdx < int(math.Min(4, float64(len(ipStr))-2)) {
		first = ipStr[0:firstIdx]
		isValid := isValidPart(first)
		if !isValid {
			break
		}

		secondIdx = firstIdx + 1
		thirdIdx = secondIdx + 1
		for secondIdx < int(math.Min(7, float64(len(ipStr))-1)) {
			second = ipStr[firstIdx:secondIdx]
			isValid = isValidPart(second)
			if !isValid {
				break
			}

			thirdIdx = secondIdx + 1
			for thirdIdx < len(ipStr) {
				third = ipStr[secondIdx:thirdIdx]
				fourth = ipStr[thirdIdx:]

				if isValidPart(third) && isValidPart(fourth) {
					currentIP = fmt.Sprintf("%s.%s.%s.%s", first, second, third, fourth)
					// fmt.Println(currentIP)
					ipList = append(ipList, currentIP)
				}
				thirdIdx++
			}
			secondIdx++
		}
		firstIdx++
	}
	return ipList
}
